### PR TITLE
Add a lockfile for frontend dependencies

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

We need to make our Docker build reproducible. This requires to add a dependencies lockfile in the project.

## Proposal

- [x] add `package-lock.json`

During this work, we also:

- [x] fixed plugins fake initialization
- [x] add CI quality checks

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -39,6 +35,36 @@ jobs:
...
+    docker:
+      - image: circleci/buildpack-deps:stretch
+    working_directory: ~/etherpad
+    steps:
+      # Checkout repository sources
+      - checkout
+      - run:
+          name: Check lockfile consistency
+          command: |
+            if git whatchanged --name-only --pretty="" origin..HEAD | grep package.json; then
```
Comment (lunika) [inline .circleci/config.yml:64]: IMO this is wrong, if you change something else than `dependencies` or `devDependencies` the lock will not be updated and the CI will fail.

File: Dockerfile [hunk 2]
```diff
@@ -24,20 +24,20 @@ RUN tar xzf epl.tgz --strip-components=1
 
 ENV NODE_ENV=production
 RUN cd src && \
-      npm install
+      npm install --no-progress --no-audit
 
 # Install extra plugins
-COPY package.json package.json
+COPY package.json package-lock.json /builder/
```
Comment (lunika) [inline Dockerfile:30]: `source` should be not a comma separated list ?

File: Dockerfile [hunk 3]
```diff
@@ -24,20 +24,20 @@ RUN tar xzf epl.tgz --strip-components=1
...
 COPY src/plugins /builder/src/plugins/
-RUN npm install
+RUN npm install --no-progress --no-audit
```
Comment (lunika) [inline Dockerfile:32]: `--no-progress` is still a performance improvement ?

File: Makefile [hunk 4]
```diff
@@ -14,6 +14,11 @@ WAIT_GRAYLOG = $(COMPOSE_RUN) dockerize -wait tcp://graylog:9000 -timeout 120s
 default: help
 
 # -- Project
+package-lock.json: package.json
+	@echo "Updating package-lock.json file..."
+	@$(COMPOSE_RUN) -T --no-deps etherpad \
+	  cat package-lock.json > package-lock.json
```
Comment (lunika) [inline Makefile:20]: this is the first time I see this :grimacing: it's a little bit strange... but why not.

File: Dockerfile (hunk 3 above)
Comment (jmaupetit) [reply Dockerfile:32]: No idea, but this prevent breaking the build logs.

File: Makefile (hunk 4 above)
Comment (jmaupetit) [reply Makefile:20]: How would you handle it? This avoids having to build locally (outside from Docker) to update the file.

File: Dockerfile (hunk 2 above)
Comment (jmaupetit) [reply Dockerfile:30]: Nope.

> Multiple <src> resources may be specified but the paths of files and directories will be interpreted as relative to the source of the context of the build.

Reference: https://docs.docker.com/engine/reference/builder/#copy

File: .circleci/config.yml (hunk 1 above)
Comment (jmaupetit) [reply .circleci/config.yml:64]: Awwwh. You are totally right. Even if in 95% cases, we will only change/update (dev)dependencies...

I hesitated while writing this part, but I think the best way to ensure both `package.json` and related lock file are synchronous is to switch to Yarn and use the `--frozen-lockfile` mode. I'll handle it in a separate PR if you don't mind?

Comment (lunika) [reply .circleci/config.yml:64]: Ok but you remove this check in this PR ?

File: Makefile (hunk 4 above)
Comment (lunika) [reply Makefile:20]: I don't know, this looks like a little bit shaky and error prone. It's not possible to use this project without this makefile

Comment (jmaupetit) [reply Makefile:20]: Can you explain how do you proceed to update Marsha's project lock file?

Comment (lunika) [reply Makefile:20]: in richie or marsha the lock file is updated when we add a library or by the renovate bot.

**Review by lunika — APPROVED** [review]

File: .circleci/config.yml (hunk 1 above)
Comment (jmaupetit) [reply .circleci/config.yml:64]: I would prefer having false positives than to check at all. I'll make sure this job is not required.

File: Makefile (hunk 4 above)
Comment (jmaupetit) [reply Makefile:20]: Current rule avoid having to run dependency updates locally. I'm not sure to understand why you think it would be error prone.

**Pull request merged**